native/src/main/java/io/ballerina/stdlib/data/xmldata/xml/XmlTraversal.java

```diff
@@ -218,10 +249,18 @@ private BMap<BString, Object> updateNextRecord(BXmlItem xmlItem, RecordType reco
             return nextValue;
         }
 
-        private BMap<BString, Object> updateNextValue(RecordType recordType, String fieldName, Type fieldType,
+        private BMap<BString, Object> updateNextValue(Type type, String fieldName, Type fieldType,
                                        BMap<BString, Object> currentMapValue, XmlAnalyzerData analyzerData) {
             analyzerData.currentField = null;
-            BMap<BString, Object> nextValue = ValueCreator.createRecordValue(recordType);
+
+            BMap<BString, Object> nextValue;
+            switch (type.getTag()) {
+                case TypeTags.RECORD_TYPE_TAG -> nextValue = ValueCreator.createRecordValue((RecordType) type);
+                case TypeTags.MAP_TAG -> nextValue = ValueCreator.createMapValue((MapType) type);
+                case TypeTags.ANYDATA_TAG, TypeTags.JSON_TAG ->
+                        nextValue = ValueCreator.createMapValue(TypeCreator.createMapType(type));
+                default -> throw DiagnosticLog.error(DiagnosticErrorCode.UNSUPPORTED_TYPE, type);
```

Comment:
Shall we add a test with `any` type

Reply:
since the method only supports `record{|anydata...|}`, for the `any` type we get compile-time error. I will negative tests with `table`, `regexp` and `xml` as exp type